File: easy-random-bean-validation/src/test/java/org/jeasy/random/validation/DecimalMinMaxAnnotationHandlerTest.java
package org.jeasy.random.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import org.jeasy.random.EasyRandom;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

class DecimalMinMaxAnnotationHandlerTest {
  private EasyRandom easyRandom;

  @BeforeEach
  void setUp() {
    easyRandom = new EasyRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    BigDecimal bigDecimal = new BigDecimal("10.0");

    TestBean testBean = easyRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestMinBigDecimal()).isGreaterThanOrEqualTo(bigDecimal);
    assertThat(testBean.getTestMaxBigDecimal()).isLessThanOrEqualTo(bigDecimal);
    assertThat(testBean.getTestMinBigInteger()).isGreaterThanOrEqualTo(BigInteger.TEN);
    assertThat(testBean.getTestMaxBigInteger()).isLessThanOrEqualTo(BigInteger.TEN);
    assertThat(new BigDecimal(testBean.getTestMinString())).isGreaterThanOrEqualTo(bigDecimal);
    assertThat(new BigDecimal(testBean.getTestMaxString())).isLessThanOrEqualTo(bigDecimal);
    assertThat(testBean.getTestMinPrimitiveByte()).isGreaterThanOrEqualTo(Byte.valueOf("10"));
    assertThat(testBean.getTestMaxPrimitiveByte()).isLessThanOrEqualTo(Byte.valueOf("10"));
    assertThat(testBean.getTestMinWrapperByte()).isGreaterThanOrEqualTo(Byte.valueOf("10"));
    assertThat(testBean.getTestMaxWrapperByte()).isLessThanOrEqualTo(Byte.valueOf("10"));
    assertThat(testBean.getTestMinPrimitiveShort()).isGreaterThanOrEqualTo(Short.valueOf("10"));
    assertThat(testBean.getTestMaxPrimitiveShort()).isLessThanOrEqualTo(Short.valueOf("10"));
    assertThat(testBean.getTestMinWrapperShort()).isGreaterThanOrEqualTo(Short.valueOf("10"));
    assertThat(testBean.getTestMaxWrapperShort()).isLessThanOrEqualTo(Short.valueOf("10"));
    assertThat(testBean.getTestMinPrimitiveInt()).isGreaterThanOrEqualTo(Integer.valueOf("10"));
    assertThat(testBean.getTestMaxPrimitiveInt()).isLessThanOrEqualTo(Integer.valueOf("10"));
    assertThat(testBean.getTestMinWrapperInt()).isGreaterThanOrEqualTo(Integer.valueOf("10"));
    assertThat(testBean.getTestMaxWrapperInt()).isLessThanOrEqualTo(Integer.valueOf("10"));
    assertThat(testBean.getTestMinPrimitiveLong()).isGreaterThanOrEqualTo(Long.valueOf("10"));
    assertThat(testBean.getTestMaxPrimitiveLong()).isLessThanOrEqualTo(Long.valueOf("10"));
    assertThat(testBean.getTestMinWrapperLong()).isGreaterThanOrEqualTo(Long.valueOf("10"));
    assertThat(testBean.getTestMaxWrapperLong()).isLessThanOrEqualTo(Long.valueOf("10"));
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @DecimalMin("10.0")
    private BigDecimal testMinBigDecimal;

    @DecimalMax("10.0")
    private BigDecimal testMaxBigDecimal;

    @DecimalMin("10.0")
    private BigInteger testMinBigInteger;

    @DecimalMax("10.0")
    private BigInteger testMaxBigInteger;

    @DecimalMin("10.0")
    private String testMinString;

    @DecimalMax("10.0")
    private String testMaxString;

    @DecimalMin("10.0")
    private byte testMinPrimitiveByte;

    @DecimalMax("10.0")
    private byte testMaxPrimitiveByte;

    @DecimalMin("10.0")
    private Byte testMinWrapperByte;

    @DecimalMax("10.0")
    private Byte testMaxWrapperByte;

    @DecimalMin("10.0")
    private short testMinPrimitiveShort;

    @DecimalMax("10.0")
    private short testMaxPrimitiveShort;

    @DecimalMin("10.0")
    private Short testMinWrapperShort;

    @DecimalMax("10.0")
    private Short testMaxWrapperShort;

    @DecimalMin("10.0")
    private int testMinPrimitiveInt;

    @DecimalMax("10.0")
    private int testMaxPrimitiveInt;

    @DecimalMin("10.0")
    private Integer testMinWrapperInt;

    @DecimalMax("10.0")
    private Integer testMaxWrapperInt;

    @DecimalMin("10.0")
    private long testMinPrimitiveLong;

    @DecimalMax("10.0")
    private long testMaxPrimitiveLong;

    @DecimalMin("10.0")
    private Long testMinWrapperLong;

    @DecimalMax("10.0")
    private Long testMaxWrapperLong;

    public BigInteger getTestMinBigInteger() {
      return testMinBigInteger;
    }

    public BigInteger getTestMaxBigInteger() {
      return testMaxBigInteger;
    }

    public String getTestMinString() {
      return testMinString;
    }

    public String getTestMaxString() {
      return testMaxString;
    }

    public byte getTestMinPrimitiveByte() {
      return testMinPrimitiveByte;
    }

    public byte getTestMaxPrimitiveByte() {
      return testMaxPrimitiveByte;
    }

    public Byte getTestMinWrapperByte() {
      return testMinWrapperByte;
    }

    public Byte getTestMaxWrapperByte() {
      return testMaxWrapperByte;
    }

    public short getTestMinPrimitiveShort() {
      return testMinPrimitiveShort;
    }

    public short getTestMaxPrimitiveShort() {
      return testMaxPrimitiveShort;
    }

    public Short getTestMinWrapperShort() {
      return testMinWrapperShort;
    }

    public Short getTestMaxWrapperShort() {
      return testMaxWrapperShort;
    }

    public int getTestMinPrimitiveInt() {
      return testMinPrimitiveInt;
    }

    public int getTestMaxPrimitiveInt() {
      return testMaxPrimitiveInt;
    }

    public Integer getTestMinWrapperInt() {
      return testMinWrapperInt;
    }

    public Integer getTestMaxWrapperInt() {
      return testMaxWrapperInt;
    }

    public long getTestMinPrimitiveLong() {
      return testMinPrimitiveLong;
    }

    public long getTestMaxPrimitiveLong() {
      return testMaxPrimitiveLong;
    }

    public Long getTestMinWrapperLong() {
      return testMinWrapperLong;
    }

    public Long getTestMaxWrapperLong() {
      return testMaxWrapperLong;
    }

    public BigDecimal getTestMinBigDecimal() {
      return testMinBigDecimal;
    }

    public BigDecimal getTestMaxBigDecimal() {
      return testMaxBigDecimal;
    }
  }
}

File: bean-validation/src/test/java/io/github/krandom/validation/DecimalMinMaxAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeGreaterThanOrEqualTo
import io.kotest.matchers.comparables.shouldBeLessThanOrEqualTo
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.DecimalMax
import jakarta.validation.constraints.DecimalMin
import java.math.BigDecimal
import java.math.BigInteger
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class DecimalMinMaxAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val bigDecimal = BigDecimal("10.0")
    val byte = 10.toByte()
    val short = 10.toShort()

    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testMinBigDecimal shouldBeGreaterThanOrEqualTo bigDecimal
    testBean.testMaxBigDecimal shouldBeLessThanOrEqualTo bigDecimal
    testBean.testMinBigInteger shouldBeGreaterThanOrEqualTo BigInteger.TEN
    testBean.testMaxBigInteger shouldBeLessThanOrEqualTo BigInteger.TEN
    BigDecimal(testBean.testMinString) shouldBeGreaterThanOrEqualTo bigDecimal
    BigDecimal(testBean.testMaxString) shouldBeLessThanOrEqualTo bigDecimal
    testBean.testMinPrimitiveByte shouldBeGreaterThanOrEqualTo byte
    testBean.testMaxPrimitiveByte shouldBeLessThanOrEqualTo byte
    testBean.testMinWrapperByte shouldBeGreaterThanOrEqualTo byte
    testBean.testMaxWrapperByte shouldBeLessThanOrEqualTo byte
    testBean.testMinPrimitiveShort shouldBeGreaterThanOrEqualTo short
    testBean.testMaxPrimitiveShort shouldBeLessThanOrEqualTo short
    testBean.testMinWrapperShort shouldBeGreaterThanOrEqualTo short
    testBean.testMaxWrapperShort shouldBeLessThanOrEqualTo short
    testBean.testMinPrimitiveInt shouldBeGreaterThanOrEqualTo 10
    testBean.testMaxPrimitiveInt shouldBeLessThanOrEqualTo 10
    testBean.testMinWrapperInt shouldBeGreaterThanOrEqualTo 10
    testBean.testMaxWrapperInt shouldBeLessThanOrEqualTo 10
    testBean.testMinPrimitiveLong shouldBeGreaterThanOrEqualTo 10L
    testBean.testMaxPrimitiveLong shouldBeLessThanOrEqualTo 10L
    testBean.testMinWrapperLong shouldBeGreaterThanOrEqualTo 10L
    testBean.testMaxWrapperLong shouldBeLessThanOrEqualTo 10L
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:DecimalMin("10.0") val testMinBigDecimal: BigDecimal,
    @field:DecimalMax("10.0") val testMaxBigDecimal: BigDecimal,
    @field:DecimalMin("10.0") val testMinBigInteger: BigInteger,
    @field:DecimalMax("10.0") val testMaxBigInteger: BigInteger,
    @field:DecimalMin("10.0") val testMinString: String,
    @field:DecimalMax("10.0") val testMaxString: String,
    @field:DecimalMin("10.0") val testMinPrimitiveByte: Byte,
    @field:DecimalMax("10.0") val testMaxPrimitiveByte: Byte,
    @field:DecimalMin("10.0") val testMinWrapperByte: Byte,
    @field:DecimalMax("10.0") val testMaxWrapperByte: Byte,
    @field:DecimalMin("10.0") val testMinPrimitiveShort: Short,
    @field:DecimalMax("10.0") val testMaxPrimitiveShort: Short,
    @field:DecimalMin("10.0") val testMinWrapperShort: Short,
    @field:DecimalMax("10.0") val testMaxWrapperShort: Short,
    @field:DecimalMin("10.0") val testMinPrimitiveInt: Int,
    @field:DecimalMax("10.0") val testMaxPrimitiveInt: Int,
    @field:DecimalMin("10.0") val testMinWrapperInt: Int,
    @field:DecimalMax("10.0") val testMaxWrapperInt: Int,
    @field:DecimalMin("10.0") val testMinPrimitiveLong: Long,
    @field:DecimalMax("10.0") val testMaxPrimitiveLong: Long,
    @field:DecimalMin("10.0") val testMinWrapperLong: Long,
    @field:DecimalMax("10.0") val testMaxWrapperLong: Long,
  )
}
